python/GafferImageUI/ImageViewUI.py

```diff
@@ -1063,11 +1107,152 @@ def __stateChanged( self, imageGadget ) :
+			i.viewWidget()._qtWidget().setProperty( "gafferAdjoinedLeft", True )
+
+		self.__row[:] = widgets
+
+class _CompareModePlugValueWidget( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug, **kw ) :
+
+		self.__button = GafferUI.MenuButton(
+			image = "compareModeNone.png",
+			menu = GafferUI.Menu(
+				Gaffer.WeakMethod( self.__menuDefinition ),
+				title = "Compare Mode",
+			)
+		)
+		self.__button._qtWidget().setMaximumWidth( 25 )
+
+		GafferUI.PlugValueWidget.__init__( self, self.__button, plug, **kw )
+
+		self.__iconDict = {
+			"" : "compareModeNone.png",
+			"replace" : "compareModeReplace.png",
+			"over" : "compareModeOver.png",
+			"under" : "compareModeUnder.png",
+			"difference" : "compareModeDifference.png",
+			"sideBySide" : "compareModeSideBySide.png"
+		}
+
+		plug.node().viewportGadget().keyPressSignal().connect(
+			Gaffer.WeakMethod( self.__keyPress ),
+			scoped = False
+		)
+
+		self._updateFromPlug()
+
+	def _updateFromPlug( self ) :
+
+		with Gaffer.Context() :
+			v = self.getPlug().getValue()
+
+		if v != "":
+			Gaffer.Metadata.registerValue( self.getPlug(), "imageViewer:lastCompareMode", v )
+
+		icon = self.__iconDict[v] if v in self.__iconDict else "compareModeNone.png"
+		self.__button.setImage( icon )
+
+
+		# We only want to set gafferAdjoinedRight on ourselves when the other controls are visible.
+		# When we're showing by ourselves, we want proper corners on the right edge.
+		# WidgetAlgo.joinEdges doesn't support unjoining edges, so just set the property manually
+		self.__button._qtWidget().setProperty( "gafferAdjoinedRight", v != "" )
+		self._repolish()
+		if self.parent():
+			self.parent()[-1].setVisible( v != "" )
+
+			self.parent()[-1]._repolish()
+
+	def __hotkeyTarget( self ):
+		with Gaffer.Context() :
+			v = self.getPlug().getValue()
+
+		if v == "":
+			return Gaffer.Metadata.value( self.getPlug(), "imageViewer:lastCompareMode" ) or "over"
+		else:
+			return ""
+
+	def __keyPress( self, gadget, event ) :
+
+		if event.key == "D" :
+			self.__setValue( self.__hotkeyTarget() )
+			return True
+
+		return False
+
+	def __menuDefinition( self ) :
+
+		with self.getContext() :
+			compareMode = self.getPlug().getValue()
+
+		hotkeyTarget = self.__hotkeyTarget()
+
+		m = IECore.MenuDefinition()
+		for name, value in [
+			( "None", "" ),
+			( "Replace", "replace" ),
+			( "Over", "over" ),
+			( "Under", "under" ),
+			( "Difference", "difference" ),
+		] :
+			m.append(
+				"/" + name,
+				{
+					"command" : functools.partial( Gaffer.WeakMethod( self.__setValue ), value ),
+					"icon" : self.__iconDict[value],
+					"shortCut" : "D" if value == hotkeyTarget else None
+				}
+			)
+
+		return m
+
+	def __setValue( self, value, *unused ) :
+
+		turningOn = value and not self.getPlug().getValue()
+
+		self.getPlug().setValue( value )
+
+		node = self.getPlug().node()
+		inputImage = node._getPreprocessor()["in"]
```

Comment:
Probably we should disable the comparison modes and the wipe tool in the ShaderView anyway?

Reply:
Once we support  comparing to an arbitrary image node, I guess it could be useful to compare a shader ball to a reference image of some sort?  For the moment, I've just fixed this issue by using a better source plug.